File: app/src/main/java/com/konektedi/vs/motions/opinions/OpinionsDataSource.java
package com.konektedi.vs.motions.opinions;

import android.arch.lifecycle.MutableLiveData;
import android.arch.paging.PageKeyedDataSource;
import android.support.annotation.NonNull;

import com.konektedi.vs.utilities.api.Api;
import com.konektedi.vs.utilities.api.ApiUtilities;
import com.konektedi.vs.utilities.NetworkState;
import com.konektedi.vs.utilities.NetworkStatus;

import java.util.List;
import java.util.concurrent.Executor;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;


public class OpinionsDataSource extends PageKeyedDataSource<Integer, Opinions> {
    private static final String TAG = "OpinionsDataSource";

    private Api apiClient;
    private MutableLiveData networkState;
    private MutableLiveData initialLoading;
    private Executor retryExecutor;
    private int motion_id;

    OpinionsDataSource(Executor retryExecutor, int motion_id) {
        apiClient = ApiUtilities.getClient();
        networkState = new MutableLiveData();
        initialLoading = new MutableLiveData();
        this.retryExecutor = retryExecutor;
        this.motion_id = motion_id;
    }

    public MutableLiveData getNetworkState() {
        return networkState;
    }

    public MutableLiveData getInitialLoading() {
        return initialLoading;
    }

    @Override
    public void loadInitial(@NonNull LoadInitialParams<Integer> params,
                            @NonNull final LoadInitialCallback<Integer, Opinions> callback) {

        networkState.postValue(NetworkState.LOADING);

        Call<List<Opinions>> request = apiClient.getOpinions(motion_id, 0);

        request.enqueue(new Callback<List<Opinions>>() {
            @Override
            public void onResponse(Call<List<Opinions>> call, Response<List<Opinions>> response) {
                if (response.isSuccessful() && response.code() == 200) {
                    callback.onResult(response.body(), 0, 10);
                    initialLoading.postValue(NetworkState.LOADED);
                    networkState.postValue(NetworkState.LOADED);


                } else {

                    initialLoading.postValue(new NetworkState(NetworkStatus.FAILED, response.message()));
                    networkState.postValue(new NetworkState(NetworkStatus.FAILED, response.message()));
                }
            }

            @Override
            public void onFailure(Call<List<Opinions>> call, Throwable t) {
                String errorMessage;
                errorMessage = t.getMessage();
                if (t == null) {
                    errorMessage = "unknown error";
                }
                networkState.postValue(new NetworkState(NetworkStatus.FAILED, errorMessage));
            }
        });
    }

    @Override
    public void loadBefore(@NonNull LoadParams<Integer> params,
                           @NonNull LoadCallback<Integer, Opinions> callback) {
    }

    @Override
    public void loadAfter(@NonNull LoadParams<Integer> params,
                          @NonNull LoadCallback<Integer, Opinions> callback) {

        networkState.postValue(NetworkState.LOADING);

        Call<List<Opinions>> request = apiClient.getOpinions(motion_id, params.key);

        request.enqueue(new Callback<List<Opinions>>() {
            @Override
            public void onResponse(Call<List<Opinions>> call, Response<List<Opinions>> response) {
                if (response.isSuccessful()) {
                    callback.onResult(response.body(), params.key + 10);
                    networkState.postValue(NetworkState.LOADED);

                } else {
                    networkState.postValue(new NetworkState(NetworkStatus.FAILED, response.message()));

                }
            }

            @Override
            public void onFailure(Call<List<Opinions>> call, Throwable t) {
                String errorMessage;
                errorMessage = t.getMessage();
                if (t == null) {
                    errorMessage = "unknown error";
                }
                networkState.postValue(new NetworkState(NetworkStatus.FAILED, errorMessage));
            }
        });
    }
}

File: app/src/main/java/com/konektedi/vs/opinions/OpinionsDataSource.kt
package com.konektedi.vs.opinions

import androidx.lifecycle.MutableLiveData
import androidx.paging.PageKeyedDataSource
import com.konektedi.vs.utilities.common.NetworkState
import com.konektedi.vs.utilities.api.Api
import com.konektedi.vs.utilities.models.Opinion
import retrofit2.Call
import retrofit2.Response
import java.io.IOException
import java.util.concurrent.Executor

class OpinionsDataSource(private val retryExecutor: Executor, private val motionId: Int) : 
        PageKeyedDataSource<Int, Opinion>() {

    // keep a function reference for the retry event
    private var retry: (() -> Any)? = null
    private val apiClient = Api.create()

    /**
     * There is no sync on the state because paging will always call loadInitial first then wait
     * for it to return some success value before calling loadAfter.
     */
    val networkState = MutableLiveData<NetworkState>()

    val initialLoad = MutableLiveData<NetworkState>()

    fun retryAllFailed() {
        val prevRetry = retry
        retry = null
        prevRetry?.let {
            retryExecutor.execute {
                it.invoke()
            }
        }
    }

    override fun loadBefore(
            params: LoadParams<Int>,
            callback: LoadCallback<Int, Opinion>) {
        // ignored, since we only ever append to our initial load
    }

    override fun loadAfter(params: LoadParams<Int>, callback: LoadCallback<Int, Opinion>) {
        networkState.postValue(NetworkState.LOADING)

        apiClient.getOpinions(motionId, params.key).enqueue(
                object : retrofit2.Callback<List<Opinion>> {
                    override fun onFailure(call: Call<List<Opinion>>, t: Throwable) {
                        retry = {
                            loadAfter(params, callback)
                        }
                        networkState.postValue(NetworkState.error(t.message ?: "unknown err"))
                    }

                    override fun onResponse(
                            call: Call<List<Opinion>>,
                            response: Response<List<Opinion>>) {
                        when {
                            response.isSuccessful -> {
                                retry = null
                                callback.onResult(response.body() as MutableList<Opinion>, params.key + 10)
                                networkState.postValue(NetworkState.LOADED)
                            }
                            response.code() == 404 -> networkState.postValue(NetworkState.END)
                            else -> {
                                retry = {
                                    loadAfter(params, callback)
                                }
                                networkState.postValue(
                                        NetworkState.error("error code: ${response.code()}"))
                            }
                        }
                    }
                }
        )
    }

    override fun loadInitial(
            params: LoadInitialParams<Int>,
            callback: LoadInitialCallback<Int, Opinion>) {

        val request = apiClient.getOpinions(motionId, 0)
        initialLoad.postValue(NetworkState.LOADING)

        // triggered by a refresh, we better execute sync
        try {
            retry = null
            val response = request.execute()

            when {
                response.isSuccessful -> {
                    networkState.postValue(NetworkState.LOADED)
                    initialLoad.postValue(NetworkState.LOADED)
                    callback.onResult(response.body() as MutableList<Opinion>, 0, 10)
                }
                response.code() == 404 -> networkState.postValue(NetworkState.END)
                else -> networkState.postValue(
                        NetworkState.error("error code: ${response.code()}"))
            }

        } catch (ioException: IOException) {
            retry = {
                loadInitial(params, callback)
            }
            val error = NetworkState.error(ioException.message ?: "unknown error")
            networkState.postValue(error)
            initialLoad.postValue(error)
        }
    }
}